File: front/src/types/achievement.ts
export type AchievementCategory = 'savings' | 'streak' | 'expense' | 'special'

export type AchievementTier = 'bronze' | 'silver' | 'gold' | 'platinum'

// カワウソの成長ステージ（none は装備なし）。閾値はバックエンド（Achievement::GROWTH_STAGES）が持つ
export const OTTER_GROWTH_STAGES = ['none', 'bronze', 'silver', 'gold', 'platinum'] as const

export type OtterGrowthStage = (typeof OTTER_GROWTH_STAGES)[number]

// 取引登録・更新レスポンスに含まれる新規解除実績
export interface ApiNewlyUnlockedAchievement {
  id: number
  title: string
  description: string
  tier: AchievementTier
  category: AchievementCategory
  reward: string
  image_url: string | null
}

// ========== API 型（スネークケース） ==========

export interface ApiAchievement {
  id: number
  original_achievement_id: string
  title: string
  description: string
  category: AchievementCategory
  unlocked: boolean
  progress: number
  progress_percentage: number
  progress_target: number
  image_url: string | null
  reward: string
  tier: AchievementTier
  created_at: string
  updated_at: string
  unlocked_at: string | null
}

export interface ApiGrowthStage {
  stage: OtterGrowthStage
  next_stage: OtterGrowthStage | null
  achievements_to_next: number | null
  progress_percentage: number
}

export interface ApiAchievementSummary {
  total_achievements: number
  unlocked_achievements: number
  // 旧バックエンドのレスポンスには存在しないため optional
  growth_stage?: ApiGrowthStage
  progress_by_category: {
    [key in AchievementCategory]?: {
      total: number
      unlocked: number
      progress_percentage: number
    }
  }
}

export interface AchievementResponse {
  achievements: ApiAchievement[]
  summary: ApiAchievementSummary
}

// ========== 内部型（キャメルケース） ==========

// 取引登録・更新で新たに解除された実績
export interface NewlyUnlockedAchievement {
  id: number
  title: string
  description: string
  tier: AchievementTier
  category: AchievementCategory
  reward: string
  imageUrl: string | null
}

export interface Achievement {
  id: number
  originalAchievementId: string
  title: string
  description: string
  category: AchievementCategory
  unlocked: boolean
  progress: number
  progressPercentage: number
  progressTarget: number
  imageUrl: string | null
  reward: string
  tier: AchievementTier
  createdAt: string
  updatedAt: string
  unlockedAt: string | null
}

export interface GrowthStage {
  stage: OtterGrowthStage
  nextStage: OtterGrowthStage | null
  achievementsToNext: number | null
  progressPercentage: number
}

export interface AchievementSummary {
  totalAchievements: number
  unlockedAchievements: number
  growthStage: GrowthStage
  progressByCategory: {
    [key in AchievementCategory]?: {
      total: number
      unlocked: number
      progressPercentage: number
    }
  }
}

// ========== マッピング関数 ==========

export function mapApiNewlyUnlockedAchievement(a: ApiNewlyUnlockedAchievement): NewlyUnlockedAchievement {
  return {
    id: a.id,
    title: a.title,
    description: a.description,
    tier: a.tier,
    category: a.category,
    reward: a.reward,
    imageUrl: a.image_url,
  }
}

export function mapApiAchievement(a: ApiAchievement): Achievement {
  return {
    id: a.id,
    originalAchievementId: a.original_achievement_id,
    title: a.title,
    description: a.description,
    category: a.category,
    unlocked: a.unlocked,
    progress: a.progress,
    progressPercentage: a.progress_percentage,
    progressTarget: a.progress_target,
    imageUrl: a.image_url,
    reward: a.reward,
    tier: a.tier,
    createdAt: a.created_at,
    updatedAt: a.updated_at,
    unlockedAt: a.unlocked_at,
  }
}

const DEFAULT_GROWTH_STAGE: GrowthStage = {
  stage: 'none',
  nextStage: null,
  achievementsToNext: null,
  progressPercentage: 0,
}

function isOtterGrowthStage(v: unknown): v is OtterGrowthStage {
  return typeof v === 'string' && (OTTER_GROWTH_STAGES as readonly string[]).includes(v)
}

// 外部データ（API レスポンス）を検証して成長ステージへ変換する。不正・欠落時は none にフォールバックする
export function parseGrowthStage(v: unknown): GrowthStage {
  if (typeof v !== 'object' || v === null || Array.isArray(v)) return DEFAULT_GROWTH_STAGE
  const raw = v as Record<string, unknown>
  if (!isOtterGrowthStage(raw.stage)) return DEFAULT_GROWTH_STAGE

  return {
    stage: raw.stage,
    nextStage: isOtterGrowthStage(raw.next_stage) ? raw.next_stage : null,
    achievementsToNext: typeof raw.achievements_to_next === 'number' ? raw.achievements_to_next : null,
    progressPercentage: typeof raw.progress_percentage === 'number' ? raw.progress_percentage : 0,
  }
}

export function mapApiAchievementSummary(s: ApiAchievementSummary): AchievementSummary {
  const progressByCategory = Object.fromEntries(
    Object.entries(s.progress_by_category).map(([key, val]) => [
      key,
      val
        ? { total: val.total, unlocked: val.unlocked, progressPercentage: val.progress_percentage }
        : undefined,
    ])
  ) as AchievementSummary['progressByCategory']

  return {
    totalAchievements: s.total_achievements,
    unlockedAchievements: s.unlocked_achievements,
    growthStage: parseGrowthStage(s.growth_stage),
    progressByCategory,
  }
}

File: back/app/models/achievement.rb
# frozen_string_literal: true

class Achievement < ApplicationRecord
  belongs_to :user

  # 実績の種類を定義
  enum :category, {
    savings: 0,      # 貯金関連
    streak: 1,       # 連続記録
    expense: 2,      # マイルストーン
    special: 3       # 特別な実績
  }

  # 実績のティア（難易度）を定義
  enum :tier, {
    bronze: 0,
    silver: 1,
    gold: 2,
    platinum: 3
  }

  # カワウソの成長ステージ（[ステージ名, 必要な解除実績数の下限]）。閾値はここだけで管理する
  GROWTH_STAGES = [['none', 0], ['bronze', 3], ['silver', 8], ['gold', 13], ['platinum', 18]].freeze

  validates :title, presence: true
  validates :description, presence: true
  validates :category, presence: true
  validates :tier, presence: true
  validates :progress_target, presence: true, numericality: { greater_than: 0 }
  validates :original_achievement_id, presence: true, uniqueness: { scope: :user_id }

  # 実績の進捗を更新するメソッド
  def update_progress(value)
    self.progress = [value, progress_target].min
    self.unlocked = (progress >= progress_target)
    self.unlocked_at = Time.current if unlocked && unlocked_at.nil?
    save!
  end

  # 実績の進捗率を計算
  def progress_percentage
    return 0 if progress_target.zero?

    ((progress.to_f / progress_target) * 100).round(1)
  end

  # 解除済み実績数から成長ステージと次のステージまでの進捗を返す
  def self.growth_stage_for(unlocked_count)
    count = [unlocked_count.to_i, 0].max
    index = GROWTH_STAGES.rindex { |_, min| count >= min }
    stage, lower = GROWTH_STAGES[index]
    next_stage, upper = GROWTH_STAGES[index + 1]

    # 最終ステージは次がないため nil / 100% とする
    return { stage: stage, next_stage: nil, achievements_to_next: nil, progress_percentage: 100 } unless next_stage

    {
      stage: stage,
      next_stage: next_stage,
      achievements_to_next: upper - count,
      progress_percentage: ((count - lower).to_f / (upper - lower) * 100).round
    }
  end
end
